Data/HashMap/Base.hs

```diff
@@ -1728,7 +1728,27 @@ fromList = L.foldl' (\ m (k, v) -> unsafeInsert k v m) empty
 {-# INLINABLE fromList #-}
 
 -- | /O(n*log n)/ Construct a map from a list of elements.  Uses
--- the provided function to merge duplicate entries.
+-- the provided function @f@ to merge duplicate entries (@f newVal oldVal@).
+--
+-- For example:
+--
+-- > fromListWith (+) [ (x, 1) | x <- xs ]
+--
+-- will create a map with number of occurrences of each element in @xs@.
+--
+-- > fromListWith (++) [ (k, [v]) | (k, v) <- xs ]
+--
+-- will group all values by their keys in a list @xs :: [(k, v)]@ and
+-- return a @HashMap k [v]@.
+--
+-- The following example illustrates how duplicate entries are accumulated;
+-- this matters when @f@ is not commutative or not associative.
+--
+-- > fromListWith f [(k, a), (k, b), (k, c), (k, d)] = fromList [(k, f d (f c (f b a)))]
+--
+-- If @f = (++)@ as in the previous example, the lists in the resulting map
+-- contain elements in reverse order from their occurences in the original
+-- list.
```

Comment:
I think a concrete, `doctest-style` example might help get the point across more quickly.

Compare the [corresponding docs for `Data.Map`](http://hackage.haskell.org/package/containers-0.6.2.1/docs/Data-Map-Strict.html#v:fromListWith). Although these suffer from the repeated use of the `'a'` and `'b'` values, which make the issue less obvious.

Reply:
Did you mean showing the output of that expression? It is there on the right, but this is maybe not as clear and visible as it could be.

Reply:
I think `fromListWith f [(k, a), (k, b), (k, c), (k, d)] = fromList [(k, f d (f c (f b a)))]` is a bit abstract. A concrete example might make the issue a bit clearer.

Reply:
Oh yes that's true.

Reply:
Let me do this in this PR too!

Reply:
Done!